src/process.cpp

```diff
@@ -439,28 +502,49 @@ std::optional<proc::proc_t> parse(const std::string &file_name) {
   try {
     pt::read_json(file_name, tree);
 
-    auto &apps_node = tree.get_child("apps"s);
-    auto &env_vars  = tree.get_child("env"s);
+    auto &apps_node            = tree.get_child("apps"s);
+    auto &env_vars             = tree.get_child("env"s);
+    auto global_prep_nodes_opt = tree.get_child_optional("global-prep-cmd"s);
 
     auto this_env = boost::this_process::environment();
 
     for(auto &[name, val] : env_vars) {
       this_env[name] = parse_env_val(this_env, val.get_value<std::string>());
     }
 
+    
+    std::vector<proc::cmd_t> global_prep_cmds;
+    if(global_prep_nodes_opt) {
+      auto &global_prep_nodes = *global_prep_nodes_opt;
+
+      global_prep_cmds.reserve(global_prep_nodes.size());
+      for(auto &[_, prep_node] : global_prep_nodes) {
+        auto do_cmd   = parse_env_val(this_env, prep_node.get<std::string>("do"s));
+        auto undo_cmd = prep_node.get_optional<std::string>("undo"s);
+
+        if(undo_cmd) {
+          global_prep_cmds.emplace_back(std::move(do_cmd), parse_env_val(this_env, *undo_cmd));
+        }
+        else {
+          global_prep_cmds.emplace_back(std::move(do_cmd));
+        }
+      }
+    }
+
     std::set<std::string> ids;
     std::vector<proc::ctx_t> apps;
     int i = 0;
     for(auto &[_, app_node] : apps_node) {
       proc::ctx_t ctx;
 
-      auto prep_nodes_opt     = app_node.get_child_optional("prep-cmd"s);
-      auto detached_nodes_opt = app_node.get_child_optional("detached"s);
-      auto output             = app_node.get_optional<std::string>("output"s);
-      auto name               = parse_env_val(this_env, app_node.get<std::string>("name"s));
-      auto cmd                = app_node.get_optional<std::string>("cmd"s);
-      auto image_path         = app_node.get_optional<std::string>("image-path"s);
-      auto working_dir        = app_node.get_optional<std::string>("working-dir"s);
+      auto prep_nodes_opt      = app_node.get_child_optional("prep-cmd"s);
```

Comment:
You know now looking over this, we can probably just refactor this section entirely instead of spamming null checks for every variable. Nothing wrong with what you did here though, just brainstorming some ideas.

Reply:
Agreed. It does look like #584 does some significant refactoring here.

Reply:
That PR still needs a lot of work, so I wouldn't base any changes on what's in that to be honest.

For one it completely removes the UI from Sunshine which is a no go.